sklearn/utils/tests/test_fixes.py

```diff
@@ -46,3 +49,37 @@ def test_loguniform(low, high, base):
     assert loguniform(base**low, base**high).rvs(random_state=0) == loguniform(
         base**low, base**high
     ).rvs(random_state=0)
+
+
+def test_delayed_fetching_right_config():
+    """Check that `_delayed` function fetches the right config associated to
+    the main thread.
+
+    Non-regression test for:
+    https://github.com/scikit-learn/scikit-learn/issues/25239
+    """
+
+    def get_working_memory():
+        return sklearn.get_config()["working_memory"]
+
+    n_iter = 10
+
+    # by default, we register the main thread and we should retrieve the
+    # parameters defined within the context manager
+    with sklearn.config_context(working_memory=123):
+        results = Parallel(n_jobs=2, pre_dispatch=4)(
+            _delayed(get_working_memory)() for _ in range(n_iter)
+        )
+
+    assert results == [123] * n_iter
+
+    # simulate that we refer to another thread
+    local_thread = threading.Thread(target=sklearn.get_config)
+    local_thread.start()
+    local_thread.join()
+    with sklearn.config_context(working_memory=123):
+        results = Parallel(n_jobs=2, pre_dispatch=4)(
+            _delayed(get_working_memory, thread=local_thread)() for _ in range(n_iter)
+        )
+
+    assert results == [get_working_memory()] * n_iter
```

Comment:
If the default `working_memory` on the `main` thread were to change this test would fail:

```python
sklearn.set_config(working_memory=140)
# the following fails
assert results == [get_working_memory()] * n_iter
```

The less fragile assertion would be check that the `local_thread` uses the default global config:

```python
from sklearn._config import _global_config_default
assert results == [_global_config_default["working_memory"]] * n_iter
```

Reply:
Reviewing this test, it seems that I have found a bug in the way the default value of the thread argument of delayed is defined. I am working on a PR against this PR.